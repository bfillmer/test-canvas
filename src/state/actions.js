// @flow
import {createAction} from 'redux-actions'

// REDUX TYPES
export const types = {
  DRAG_START: 'svg/DRAG_START',
  DRAG_END: 'svg/DRAG_END',
  DRAGGED: 'svg/DRAGGED',
  ZOOMED: 'svg/ZOOMED'
}

// FLOW TYPES
type DragStartAction = {
  type: string,
  payload: {
    x: number,
    y: number,
    id: string
  }
}

type DragEndAction = {
  type: string
}

type DraggedAction = {
  type: string,
  payload: {
    x: number,
    y: number
  }
}

type ZoomedAction = {
  type: string,
  payload: {
    dy: number,
    id: string
  }
}

const startDrag: DragStartAction = createAction(types.DRAG_START, (x: number, y: number, id: string) => ({x, y, id}))
const endDrag: DragEndAction = createAction(types.DRAG_END)
const dragging: DraggedAction = createAction(types.DRAGGED, (x: number, y: number) => ({x, y}))
const zooming: ZoomedAction = createAction(types.ZOOMED, (dy: number, id: string) => ({dy, id}))

// ACTIONS
export const actions = {
  startDrag,
  endDrag,
  dragging,
  zooming
}
