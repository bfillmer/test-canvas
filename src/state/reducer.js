
import {createAction, handleActions} from 'redux-actions'

const assign = Object.assign

const svgObject = (
  id,
  matrix = [1, 0, 0, 1, 0, 0],
  dragging = false,
  dragX = 0,
  dragY = 0
) => ({
  id,
  dragging,
  matrix,
  dragX,
  dragY
})

// TYPES
const DRAG_START = 'svg/DRAG_START'
const DRAG_END = 'svg/DRAG_END'
const DRAGGED = 'svg/DRAGGED'
const ZOOMED = 'svg/ZOOMED'

// ACTIONS
export const actions = {
  startDrag: createAction(DRAG_START, (x, y, id) => ({x, y, id})),
  endDrag: createAction(DRAG_END),
  dragging: createAction(DRAGGED, (x, y) => ({x, y})),
  zooming: createAction(ZOOMED, (dy, id) => ({dy, id}))
}

// REDUCERS
const initialState = {
  dragging: false,
  svg: [
    svgObject('viewport'),
    svgObject('circle-one', [1, 0, 0, 1, -10, -10]),
    svgObject('circle-two', [1, 0, 0, 1, 20, 20])
  ]
}

// Update x, y in our matrix array with the delta x/y from moving.
const pan = (dx, dy, matrix) => matrix.map((n, i) => (i === 4) ? n + dx : (i === 5) ? n + dy : n)

// Soom in or out based on the delta of the wheel scroll.
const zoom = (dy, matrix) => {
  const scale = dy < 0 ? 1.05 : 0.95
  return matrix.map((n, i) => n * scale)
}

export const reducer = handleActions({
  [DRAG_START]: (state, action) => {
    const {id, x, y} = action.payload
    return assign({}, state, {
      dragging: id,
      svg: state.svg.map(svg => svg.id === id ? assign({}, svg, {
        dragX: x,
        dragY: y
      }) : svg)
    })
  },
  [DRAGGED]: (state, action) => {
    const {x, y} = action.payload
    return assign({}, state, {
      svg: state.svg.map(svg => {
        if (svg.id !== state.dragging) return svg
        const dx = x - svg.dragX
        const dy = y - svg.dragY
        return assign({}, svg, {
          matrix: pan(dx, dy, svg.matrix),
          dragX: x,
          dragY: y
        })
      })
    })
  },
  [DRAG_END]: (state, action) => assign({}, state, {
    dragging: false
  }),
  [ZOOMED]: (state, action) => assign({}, state, {
    svg: state.svg.map(svg => svg.id === action.payload.id ? assign({}, svg, { matrix: zoom(action.payload.dy, svg.matrix) }) : svg)
  })
}, initialState)
