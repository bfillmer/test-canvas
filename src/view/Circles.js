
import React from 'react'
import {connect} from 'react-redux'

const mapStateToProps = state => ({
  circles: state.svg.filter(s => s.id !== 'viewport')
})

const Container = ({circles}) => {
  const renderCircles = circles.map(s => <circle key={s.id} id={s.id} transform={`matrix(${s.matrix.join(' ')})`} r='20' fill='teal' stroke='black' />)
  return (<g>{renderCircles}</g>)
}

export const Circles = connect(mapStateToProps)(Container)
