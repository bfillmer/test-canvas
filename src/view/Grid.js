
import React from 'react'
import {connect} from 'react-redux'

const mapStateToProps = state => ({
  scalar: state.scalar,
  viewportMatrix: state.svg.filter(svg => svg.id === 'viewport').reduce((acc, val) => val.matrix, [])
})

// Viewport constants. Need to be set and leveraged both in Grid and Viewport.
const baseWidth = 1440
const baseHeight = 960
const viewboxWidth = baseWidth
const viewboxHeight = baseHeight
const halfWidth = viewboxWidth / 2
const halfHeight = viewboxHeight / 2

// @NOTE First pass code. More hardcoded into the render function than necessary.
// @TODO Set limits on scalar in the reducer to prevent excessive zoom out.
const Container = ({scalar, viewportMatrix}) => {
  // Current Center Value
  const [centerX, centerY] = viewportMatrix.slice(4, 6)
  const gridSize = 50

  // Additional grid unit added to each bound to ensure grid scales to just outside the viewport.
  // Grid unit adjusted for scalar when added, then the bould is adjusted for scalar after.
  const bounds = {
    left: (-(centerX + halfWidth) - (gridSize * scalar)) / scalar,
    right: ((halfWidth - centerX) + (gridSize * scalar)) / scalar,
    top: (-(centerY + halfHeight) - (gridSize * scalar)) / scalar,
    bottom: ((halfHeight - centerY) + (gridSize * scalar)) / scalar
  }

  // Calculate the maximum lines from the center point to any given bound direction.
  const maxLines = b => Math.ceil(b / gridSize) * gridSize

  const left = maxLines(bounds.left)
  const right = maxLines(bounds.right)
  const top = maxLines(bounds.top)
  const bottom = maxLines(bounds.bottom)

  const lines = []

  for (let n = left; n <= right; n += gridSize) {
    lines.push(<line key={`${n}-v`} x1={n} x2={n} y1={top} y2={bottom} stroke='#fff' fill='none' strokeWidth='2' />)
  }

  for (let n = top; n <= bottom; n += gridSize) {
    lines.push(<line key={`${n}-h`} x1={left} x2={right} y1={n} y2={n} stroke='#fff' fill='none' strokeWidth='2' />)
  }

  const center = (<circle cx='0' cy='0' r='6' stroke='#ddd' strokeWidth='1' fill='#ddd' />)

  // id set to viewport so that drag events fire and update the viewport matrix.
  return (
    <g id='viewport' transform={`matrix(${[1, 0, 0, 1, 0, 0].join(' ')})`}>
      {lines}
      {center}
    </g>
  )
}

export const Grid = connect(mapStateToProps)(Container)
