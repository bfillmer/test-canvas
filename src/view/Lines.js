
import React from 'react'
import {connect} from 'react-redux'

const mapStateToProps = state => ({
  lines: state.blocks.filter(block => !!block.dependsOn).map(block => {
    const a = state.svg.reduce((array, svg) => svg.id === block.dependsOn ? svg.matrix : array, [])
    const b = state.svg.reduce((array, svg) => svg.id === block.id ? svg.matrix : array, [])
    return {
      a,
      b
    }
  })
})

const point = ([x, y]) => (`${x},${y}`)

const Container = ({lines}) => {
  const renderLines = lines.map(({a, b}, key) => {
    const [aX, aY, bX, bY] = [a[4], a[5], b[4], b[5]]
    // Calculate total movement in x & y directions. [DONE]
    // Whichever is longer, move half that direction. [DONE]
    // Then the total distance in the other direction. [DONE]
    // Then finish up the rest of the way. [DONE]
    const p1 = [aX, aY]
    const [mX, mY] = [(aX + bX) / 2, (aY + bY) / 2]
    const [diffX, diffY] = [Math.abs(bX - aX), Math.abs(bY - aY)]
    const p2 = (diffX > diffY) ? [mX, aY] : [aX, mY]
    const p3 = (diffX > diffY) ? [mX, bY] : [bX, mY]
    const p4 = [bX, bY]
    const points = [
      point(p1),
      point(p2),
      point(p3),
      point(p4)
    ]
    return <polyline key={key} fill='none' stroke='#333' strokeWidth='4' strokeLinejoin='round' points={points.join(' ')} />
  })
  return (
    <g id='viewport'>
      {renderLines}
    </g>
  )
}

export const Lines = connect(mapStateToProps)(Container)
