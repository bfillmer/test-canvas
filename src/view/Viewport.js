
import React from 'react'

export const Viewport = ({
  onDragStart,
  onDragMove,
  onDragEnd,
  onWheel,
  viewportMatrix,
  children
}) => (
  <svg
    width='100%'
    height='100%'
    id='viewport'
    onMouseDown={onDragStart}
    onMouseMove={onDragMove}
    onMouseUp={onDragEnd}
    onWheel={onWheel}
    viewBox={[-(1440 / 2), -(960 / 2), 1440, 960].join(' ')}
    preserveAspectRatio='xMidYMid slice'
  >
    <g transform={`matrix(${viewportMatrix})`}>
      {children}
    </g>
  </svg>
)
