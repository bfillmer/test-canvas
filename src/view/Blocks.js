
import React from 'react'
import {connect} from 'react-redux'

import {BLOCK} from 'state/reducers'

// @NOTE Blocks would individually connect to the store to get their normalized data, thus
// the viewport only updates when anything with positioning changes, and the internal Block
// state only updates when relevant Block data changes.
const Block = ({id, matrix}) => (
  <g transform={`matrix(${matrix.join(' ')})`}>
    <rect id={id} width='150' height='100' rx='6' ry='6' fill='#ccc' stroke='#666' strokeWidth='2' x='-75' y='-50' />
  </g>
)

const mapStateToProps = state => ({
  blocks: state.svg.filter(s => s.type === BLOCK)
})

const Container = ({blocks}) => {
  const renderBlocks = blocks.map(s => <Block key={s.id} {...s} />)
  return (<g>{renderBlocks}</g>)
}

export const Blocks = connect(mapStateToProps)(Container)
