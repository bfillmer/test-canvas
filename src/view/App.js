
import React from 'react'

import {Viewport} from 'view/Viewport'
import {Blocks} from 'view/Blocks'
import {Grid} from 'view/Grid'

export const App = () => (
  <Viewport>
    <Grid />
    <Blocks />
  </Viewport>
)
