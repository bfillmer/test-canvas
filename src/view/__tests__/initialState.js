/* global test, expect */

import React from 'react'
import renderer from 'react-test-renderer'
import {createStore} from 'redux'
import {Provider} from 'react-redux'

import {reducer} from 'state/reducer'

import {App} from 'view/App'

// Import the expected "snapshot" of the redux store that describes
// our UI we wish to test.
import data from './initialState.json'

const testStore = createStore(
  reducer,
  JSON.parse(data.preloadedState)
)

test('loads initialState and renders', () => {
  const tree = renderer.create(
    <Provider store={testStore}>
      <App />
    </Provider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
