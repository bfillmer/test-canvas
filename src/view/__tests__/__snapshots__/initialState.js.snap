exports[`test loads initialState and renders 1`] = `
<svg
  height="100%"
  id="viewport"
  onMouseDown={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
  onWheel={[Function]}
  preserveAspectRatio="xMidYMid slice"
  viewBox="-720 -480 1440 960"
  width="100%">
  <g
    transform="matrix(1 0 0 1 0 0)">
    <circle
      fill="teal"
      id="circle-one"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 -10 -10)" />
    <circle
      fill="teal"
      id="circle-two"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 -10 40)" />
    <circle
      fill="teal"
      id="circle-three"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 -10 -60)" />
    <circle
      fill="teal"
      id="circle-four"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 80 -10)" />
    <circle
      fill="teal"
      id="circle-five"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 80 40)" />
    <circle
      fill="teal"
      id="circle-six"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 80 -60)" />
    <circle
      fill="teal"
      id="circle-seven"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 35 -10)" />
    <circle
      fill="teal"
      id="circle-eight"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 170 -10)" />
    <circle
      fill="teal"
      id="circle-nine"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 170 40)" />
    <circle
      fill="teal"
      id="circle-ten"
      r="20"
      stroke="black"
      transform="matrix(1 0 0 1 170 -60)" />
  </g>
</svg>
`;
